### ISSUE TITLE
scatter_ op after permute op not working correctly on mps device

### ISSUE THREAD

**Issue by pratkpranav** [pytorch/pytorch#115152]
### 🐛 Describe the bug

When I try to use scatter_ op after permute op, The resulting matrix is always zero. Following is the script for reproduction

```python

import torch

device = torch.device("mps")


def test_topk_selection():
    random_matrix = torch.randn((8, 200, 100), device=device)
    random_matrix = torch.permute(random_matrix, (0, 2, 1))

    ret = torch.topk(random_matrix, k=int(0.1 * random_matrix.size(2)), sorted=False)
    ret.indices.to(device)
    ret.values.to(device)

    res2 = torch.zeros_like(random_matrix, device=device)
    res2.scatter_(2, ret.indices, ret.values)

    res1 = torch.zeros_like(random_matrix, device=device)
    for i in range(ret.indices.size(0)):
        for j in range(ret.indices.size(1)):
            for k in range(len(ret.indices[i][j])):
                res1[i][j][ret.indices[i][j][k]] = ret.values[i][j][k]

    print(torch.count_nonzero(res1), torch.count_nonzero(res2))
    assert torch.all(torch.eq(res1, res2))


test_topk_selection()
```

### Versions

```

Collecting environment information...
PyTorch version: 2.2.0.dev20231128
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.10.12 (main, Jul 28 2023, 18:34:01) [Clang 14.0.3 (clang-1403.0.22.14.1)] (64-bit runtime)
Python platform: macOS-13.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] torch==2.2.0.dev20231128
[pip3] torchaudio==2.2.0.dev20231128
[pip3] torchvision==0.17.0.dev20231128
[pip3] torchviz==0.0.2
[conda] Could not collect

```

**Comment by ejnnr** [comment]
I ran into what might be the same issue in a slightly different setting, namely calling `scatter_` on certain subtensors. (It works fine on CPU, just not on MPS). See example below, everything is fine when calling `a[0].scatter_`, but neither `a[:, 0].scatter_` nor `a[1].scatter_` seem to work. Maybe it works iff the subtensor is contiguous and starts at position 0?

This happens on both `2.1.0` and `torch-2.2.0.dev20231206` (haven't tried any other versions yet).

```python
import torch

print("Works:")
for device in ["cpu", "mps"]:
    a = torch.zeros((2, 2), device=device)
    a[0].scatter_(0, torch.tensor([0], device=device), 42)
    print(a)

print("\nDoesn't work on MPS:")

for device in ["cpu", "mps"]:
    a = torch.zeros((2, 2), device=device)
    a[1].scatter_(0, torch.tensor([0], device=device), 42)
    print(a)

print("\nAlso doesn't work on MPS:")
for device in ["cpu", "mps"]:
    a = torch.zeros((2, 2), device=device)
    a[:, 0].scatter_(0, torch.tensor([0], device=device), 42)
    print(a)
```
prints:
```
Works:
tensor([[42.,  0.],
        [ 0.,  0.]])
tensor([[42.,  0.],
        [ 0.,  0.]], device='mps:0')

Doesn't work on MPS:
tensor([[ 0.,  0.],
        [42.,  0.]])
tensor([[0., 0.],
        [0., 0.]], device='mps:0')

Also doesn't work on MPS:
tensor([[42.,  0.],
        [ 0.,  0.]])
tensor([[0., 0.],
        [0., 0.]], device='mps:0')
```

**Comment by johnlyzhou** [comment]
Also ran into this issue in my code, works fine on CUDA but calling it on MPS produces a tensor of all 0s

**Comment by ArvinZhuang** [comment]
I run into a similar problem, `scatter_` on different devices gives different results. Below is an example on CPU, the result is correct:
```python
import torch
logits = torch.zeros(2, 4, 6, device='cpu')
labels = torch.zeros_like(logits)
shift_labels = torch.tensor([[3, 5, 5, 2], [3, 3, 3, 3]], device='cpu')


index = shift_labels.unsqueeze(1).repeat(1, logits.size(1), 1)
mask = torch.tril(torch.ones_like(index), diagonal=0).transpose(1, 2)
labels.scatter_(2, index, mask.float())

# labels now is:
tensor([[[0., 0., 1., 1., 0., 1.],
         [0., 0., 1., 0., 0., 1.],
         [0., 0., 1., 0., 0., 1.],
         [0., 0., 1., 0., 0., 0.]],

        [[0., 0., 0., 1., 0., 0.],
         [0., 0., 0., 1., 0., 0.],
         [0., 0., 0., 1., 0., 0.],
         [0., 0., 0., 1., 0., 0.]]])

```
but run exact same code on cuda, the result is different and wrong:

```python
import torch
logits = torch.zeros(2, 4, 6, device='cuda:0')
labels = torch.zeros_like(logits)
shift_labels = torch.tensor([[3, 5, 5, 2], [3, 3, 3, 3]], device='cuda:0')


index = shift_labels.unsqueeze(1).repeat(1, logits.size(1), 1)
mask = torch.tril(torch.ones_like(index), diagonal=0).transpose(1, 2)
labels.scatter_(2, index, mask.float())

# labels now is:
tensor([[[0., 0., 1., 1., 0., 1.],
         [0., 0., 1., 0., 0., 1.],
         [0., 0., 1., 0., 0., 0.],
         [0., 0., 1., 0., 0., 0.]],

        [[0., 0., 0., 1., 0., 0.],
         [0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0.]]], device='cuda:0')
```

I found a solution for cuda devices, change `labels.scatter_(2, index, mask.float())` to `labels.scatter_reduce_(2, index, mask.float(), 'amax')`, but `scatter_reduce_` is not implemented for MPS devices so far. (My torch version is 2.2.2)
